internal/pipe/gomod/gomod.go

```diff
@@ -30,5 +35,83 @@ func (Pipe) Run(ctx *context.Context) error {
 
 	ctx.ModulePath = result
 
+	if !ctx.Config.GoMod.Proxy {
+		return pipe.Skip("gomod.proxy is disabled")
+	}
+
+	if ctx.Snapshot {
+		return pipe.ErrSnapshotEnabled
+	}
+
+	return setupProxy(ctx)
+}
+
+func setupProxy(ctx *context.Context) error {
+	template := tmpl.New(ctx)
+
+	log.Infof("proxying %s@%s", ctx.ModulePath, ctx.Git.CurrentTag)
+
+	mod, err := template.Apply(`
+module {{ .ProjectName }}
+
+require {{ .ModulePath }} {{ .Tag }}
```

Comment:
One sanity check we should also add is that the version resolved by `go get` matches the version specified. 

As I understand it, the version specified is the template value `.Tag`. Going back to the CUE example, in this case that would be `v0.3.0-beta.5` - corresponding to the tag in the repo. If you run `go get` as above, that will add a requirement to the `go.mod` file with that same version identifier, `v0.3.0-beta.5`. All good.

However in some situations, `go get` resolves a different identifier. e.g. you can `go get -d cuelang.org/go/cmd/cue@$commitsha` and it will resolve to a pseudo version. In that case, the version specified (`$commitsha`) will be different to the resolved version (`v0.3.0.beta.5`). I can't think how we would end up in this situation, but we should probably fail for now.

Reply:
Incidentally, you would get the resolved version via:

```
go list -f {{.Module.Version}} $main
```

Reply:
Currently we get `ModulePath` with `go list -m`, which seems to resolve correctly:

```
~/Developer/forks/cue master
❯ go list -m
cuelang.org/go
```

Tried it locally:

```diff
diff --git a/.goreleaser.yml b/.goreleaser.yml
index 961dbf14..09b7bbf3 100644
--- a/.goreleaser.yml
+++ b/.goreleaser.yml
@@ -4,6 +4,9 @@ before:
   hooks:
     - go mod download
 
+gomod:
+  proxy: true
+
 builds:
 - env:
   - CGO_ENABLED=0
```

and then:

```sh
~/Developer/forks/cue master*
❯ ~/Developer/goreleaser/goreleaser/goreleaser release --skip-validate --skip-publish --rm-dist --parallelism 1
   • releasing...
   • loading config file       file=.goreleaser.yml
   • loading environment variables
   • getting and validating git state
      • releasing v0.3.0-beta.7, commit cbda0d34cc67041b94a1f7d9bd2557598d9f3e49
      • pipe skipped              error=validation is disabled
   • parsing tag
   • running before hooks
      • running go mod download
   • setting defaults
      • snapshotting
      • github/gitlab/gitea releases
      • project name
      • building binaries
      • creating source archive
      • archives
      • linux packages
      • snapcraft packages
      • calculating checksums
      • signing artifacts
      • docker images
      • artifactory
      • blobs
      • homebrew tap formula
         • optimistically guessing `brew[0].install` to be `bin.install "cue"`
      • scoop manifests
      • milestones
   • snapshotting
      • pipe skipped              error=not a snapshot
   • checking ./dist
      • --rm-dist is set, cleaning it up
   • loading go mod information
      • proxying cuelang.org/go@v0.3.0-beta.7 to build cuelang.org/go/cmd/cue # NOTE HERE
   • writing effective config file
      • writing                   config=dist/config.yaml
   • generating changelog
      • writing                   changelog=dist/CHANGELOG.md
   • building binaries
      • building                  binary=/Users/carlos/Developer/forks/cue/dist/cue_darwin_amd64/cue
      • building                  binary=/Users/carlos/Developer/forks/cue/dist/cue_darwin_arm64/cue
^C      ⨯ release failed after 9.73s error=received: interrupt

```

seems like it resolved correctly 🤔 

do you have examples in which you think it wouldn't work so I can take a look?

---

about the version, agree, should probably add some sort of validation there...

Reply:
Ah, seeing your worked example was helpful. Because it raises a really interesting/important question about the context in which `goreleaser` would be run when using the proxy like this.

As we've agreed, the main reason behind using the proxy and checksum is that they act as a shared unambiguous, read-only, authenticated record of a what a given version of a module represents.

Continuing the CUE example. When building a release for the CUE project, we're suggesting that the binary assets for that release should be built according to proxy-resolved and checksum-verified modules. 

But I think we also need to consider that other assets, e.g. docs, license files etc, should _also_ come from the proxy zip file. Because if we don't, the non-binary assets could be inconsistent with the binary assets. This can, and does, happen when people move tags, under the misapprehension that "things will be fine"! The issue is that the proxy will only ever see the first revision that tag resolved to (same with the sumdb), whereas the project building via `goreleaser` and using `git` tags will see whatever is current. 

So I think this leaves us in the following situation:

* A `goreleaser` release will be run with a working directory of directory that is the root of a clone of the project that you want to release, checked out at the tag corresponding to the semantic version, e.g. `v1.1.0`, that is the target version of the release.
* Here the definition of a project coincides with that of a module; see below for a note about multi-module repos. 
* `go list -m` within the working directory of the release gives the main module associated with the release. Our assertion is that main packages within this module will be the binary assets included as part of the release (with a TBD on whether we support main packages from others modules, dependencies of this main module or otherwise). Let's refer to this main module path as `$MODULE`, e.g. `cuelang.org/go`.
* The `go.sum` of the main module will be used as a reference for module authentication (per your change).
* A temporary module will be created for building binaries from main packages that belong to this project/main module, with the main module at the specified version as a dependency of that temporary module.
* Binary artefacts will be built using that temporary module for dependency resolution.
* Other non-binary files ([example](https://github.com/cuelang/cue/blob/cbda0d34cc67041b94a1f7d9bd2557598d9f3e49/.goreleaser.yml#L31-L35)) should be resolved from within the module cache so that they are consistent with the source files used to created the binaries.

Building on the last point, additional files should be added from within the module cache, with the root path given by running the following with a working directory of the temporary module:

```
go list -m -f {{.Dir}} $MODULE
```

The only wrinkle here being that files within the module cache are `0444` and directories `0555` (before umask). i.e. they lose all permissions compared to the source repository. This would, for example, be an issue for a shell script. 

Does `goreleaser` have a phase of the release that would support adjusting those permissions?

More generally, how does that sound to you?


### Multi-module repos

To what extent does `goreleaser` support multi-module repos? Because a tag in that instance would be of the form [`gopls/v0.6.9`](https://github.com/golang/tools/releases/tag/gopls%2Fv0.6.9). In this case the working directory will need to be the subdirectory that corresponds to the tag, e.g. `gopls`. Not critical we solve this now per se, unless it influences our decisions: just don't want to box ourselves in a corner :)

Reply:
We don't have a way to change permissions yet... moreover, I think the "regular user" will not expect they need to set permissions of files upon release if they already set it on the repo itself... could be solved with better docs et al... but, not sure.

If we take this file thing to the ultimate extend, then maybe it also won't make sense to use the current goreleaser config file, as the module's might be different. At that point, it wouldn't make sense to make the proxy thing before everything and releasing from within the module itself? 🤔  (assuming we have permissions et al - which we don't)

Currently goreleaser do not support multi-modules et al... e.g., we don't have a way to to say "tags prefixed with foo/ are related to the project foo on the dir whatever/foo". Its somewhere on the roadmap, but I think its not something we need to worry right now 🤔

Reply:
> We don't have a way to change permissions yet... moreover, I think the "regular user" will not expect they need to set permissions of files upon release if they already set it on the repo itself... could be solved with better docs et al... but, not sure.

Here's one option. Take the files from the module cache (per above), but use the permissions from the repo (in a mode whereby if you fail to get/set permissions from the source/on the target you ignore the error). The worst case in this situation is that you end up with the wrong permissions (if a version tag has moved and with that move the permissions were changed relative to the original tag that resolved via the proxy), but never content. 

(Incidentally, it would be good for `goreleaser` to fail in case the tag for a version tag _has_ moved. But there does not appear to be a good way to do that today; the `GOPROXY` protocol does not give us enough information - I've asked a question about this)

> but I think its not something we need to worry right now 🤔

The only thing worth considering is whether our proposed config changes work for that setup. I think they do, but was just vocalising this point for others to also consider.

Reply:
Somebody beat me to asking about the proxy returning VCS information for a given module version: https://github.com/golang/go/issues/44742

Reply:
> Here's one option. Take the files from the module cache (per above), but use the permissions from the repo (in a mode whereby if you fail to get/set permissions from the source/on the target you ignore the error). The worst case in this situation is that you end up with the wrong permissions (if a version tag has moved and with that move the permissions were changed relative to the original tag that resolved via the proxy), but never content.

I thought about that... but then it would still be wrong if the repo has different permissions now (due to moving tags for example).

If we ignore that, we would have to change **a lot** of things... and I don't think its worth it... at least not initially...

once we add the perms stuff to the archive pipe we can look into this again I think, maybe making it explicit by exposing the module cache path as a template variable or something like that...



> The only thing worth considering is whether our proposed config changes work for that setup. I think they do, but was just vocalising this point for others to also consider.

yeah I think it would... but in any case, multi-modules would require changing the very core of how goreleaser works, so it will probably happen only a v2 or something like that, in which I will probably break compatibility with v0/v1 et al.

Reply:
BTW: I think I'll start adding some tests and maybe ship that on the next release... @myitcv do you want to give it a try as well and run it against one project (can be an example/test project)?

Reply:
>  @myitcv do you want to give it a try as well and run it against one project (can be an example/test project)?

Just to confirm, you want me to test this PR?

Reply:
> Just to confirm, you want me to test this PR?

yep, I already tested with cue, but would like to hear about your experience nevertheless :)

Reply:
> I thought about that... but then it would still be wrong if the repo has different permissions now (due to moving tags for example).

Indeed, that's the scenario I outlined as well that would be problematic. However, until either https://github.com/golang/go/issues/44742 lands, or such time as you want to introduce an explicit config-driven permissions setup, I think it's sufficient. If we make it a best-efforts step: i.e. you don't want to fail on any errors in the permissions transfer.

Reply:
problem is that it would require changing goreleaser **a lot** [1]... not sure its worth it.

[1]: basically everywhere we handle files, e.g.:
- archives
- nfpm
- docker
- snapcraft
- http upload
- release
- custom publishers
- blobs
- artifactory
- bintray

(no particular order)

its a lot of changes for something "temporary"...

Reply:
> not sure its worth it.

Yep, I'm agreeing with that point 😄

Reply:
Out of interest, how did you test CUE with this change? I'm probably missing some `goreleaser` foo here, but I can't see a way to not run in snapshot mode?

Reply:
oh, `goreleaser release --rm-dist --skip-publish` (maybe `--skip-validate` as well)

Reply:
Ah, `--skip-validate` was what I was missing. 

Looks good to me. As an added incentive to land this, we're soon going to be release `v0.3.0` of CUE. Being able to make use of this would be fantastic!

Reply:
will probably work on adding tests/more docs today/tomorrow, so it should land soon :)

Reply:
If I can be of help looking anything else over, please let me know.